dev/service-registry/service-locator/src/main/scala/com/lightbend/lagom/registry/impl/ServiceRegistryActor.scala

```diff
@@ -19,100 +19,182 @@ import scala.collection.JavaConverters._
-    val addressUri = service.uri
+  def rebuild(registry: InternalRegistry): Unit = {
+    routerFunctions = registry.reg.map { case (k, v) => (k.serviceName.name, k.portName, v._1, v._2) }.toSeq.flatMap { case (name, portname, uri, details) => serviceRouter(name, portname, uri, details) }
+    simpleRegistry = registry.serviceValues.map { case (k, v) => k.name -> v }
+    router = routerFunctions.foldLeft(PartialFunction.empty[Route, (URI, ServiceRegistryService)])(_ orElse _)
+  }
+
+  private def serviceRouter(serviceName: String, registeredPortName: Option[String], uri: URI, service: ServiceRegistryService): Seq[PartialFunction[Route, (URI, ServiceRegistryService)]] = {
     // lazy because if there's no ACLs, then there's no need to create an InetSocketAddress, and hence no need to fail
     // if the port can't be calculated.
-    lazy val address = (addressUri.getScheme, addressUri.getHost, addressUri.getPort) match {
-      case (_, null, _)        => throw new IllegalArgumentException("Cannot register a URI that doesn't have a host: " + addressUri)
-      case ("http", host, -1)  => new InetSocketAddress(host, 80)
-      case ("https", host, -1) => new InetSocketAddress(host, 443)
-      case (_, _, -1)          => throw new IllegalArgumentException("Cannot register a URI that does not specify a port: " + addressUri)
-      case (_, host, port)     => new InetSocketAddress(host, port)
-    }
-    val routerFunctions: Seq[PartialFunction[Route, InetSocketAddress]] = service.acls.asScala.map {
+    val routerFunctions: Seq[PartialFunction[Route, (URI, ServiceRegistryService)]] = service.acls.asScala.map {
       case acl =>
         acl.method().asScala -> acl.pathRegex().asScala.map(Pattern.compile)
     }.map {
       case (aclMethod, pathRegex) =>
-        val pf: PartialFunction[Route, InetSocketAddress] = {
-          case Route(method, path) if aclMethod.forall(_.name == method) && pathRegex.forall(_.matcher(path).matches()) =>
-            address
+        val pf: PartialFunction[Route, (URI, ServiceRegistryService)] = {
+          case Route(method, path, requestedPortName) if aclMethod.forall(_.name == method) && pathRegex.forall(_.matcher(path).matches()) && registeredPortName == requestedPortName =>
+            (uri, service)
         }
         pf
     }
     routerFunctions
   }
 
-  private def rebuildRouter() = {
-    routerFunctions = registry.values.flatMap(serviceRouter).toSeq
-    router = routerFunctions.foldLeft(PartialFunction.empty[Route, InetSocketAddress])(_ orElse _)
+  def warnOnAmbiguity(route: Route): Unit = {
+    if (logger.isWarnEnabled) {
+      val servingRoutes = routerFunctions.filter(_.isDefinedAt(route))
+      if (servingRoutes.size > 1) {
+        val servers = servingRoutes.map(_.apply(route))
+        logger.warn(s"Ambiguous route resolution serving route: $route. Route served by ${servers.head} but also matches ${servers.tail.mkString("[", ",", "]")}.")
+      }
+    }
+  }
+
+}
```

Comment:
I extracted the relevant code into `InternalRegistry` and `InternalRouter` so it was easier to reason and extend the code. This also allowed writing some non-actor tests.

The code on the actor is now a lot cleaner too.